### DISCUSSION TITLE
rspack是对压缩算法做了优化吗，构建产物体积比webpack的小了很多

### DISCUSSION THREAD

**Discussion by jinphic** [web-infra-dev/rspack#4690] [Q&A]
看SWC的介绍，SWC压缩算法是不如babel的，体积会大2%左右。
但是在我用analyze分析了rspack的产物、和 webpack的产物后，我发现rspack的产物比webpack的要小上很多。
我自己对比构建后的代码发现可能是以下原因：
1. babel regenerator runtime 打包进去了。
   这个是会重复打包进入的，只要有多个bundle那么就会打包多份
2. babel有很多胶水代码 
3. webpack+babel 构建后的代码中有很多文件路径和文件名。但是rspack就完全没有这些

**除此之外，是否rspack还用了更高明的方法来处理React 、tsx文件 ？？**
![image](https://github.com/web-infra-dev/rspack/assets/78671853/69b761a2-e970-4103-b21e-60e8dc15b364)
![image](https://github.com/web-infra-dev/rspack/assets/78671853/a2535140-bd36-43cd-94b5-5c292d54e515)

**Comment by jinphic** [comment]
看[生产优化](https://www.rspack.dev/zh/guide/production.html) 这简单提了一下，是因为采用了 “内置的压缩器对 JavaScript 和 CSS 代码进行压缩” 吗 ？这个体积压缩效果很显著

**Reply by hardfist** [reply]
估计你webpack构建优化没开最优，webpack prod的moduleId应该是number不是named,查看下你optimization.moduleId的配置

**Reply by jinphic** [reply]
感谢解答

**Comment by JSerFeng** [comment]
Rspack 压缩默认使用 SWC minifier，他和 Terser 的压缩率接近，有的时候 SWC minifier会更优秀一点，但不会相差很多。

对于提到的 babel regenerator runtime 和胶水（helper）代码，在同样需要语法降级的前提下的话，我们也会有 SWC 的 regenerator runtime 和 swc helper 的代码，总的来说体积应该是比较接近的，这里可以通过 splitChunks 把 helper 单独打一个 chunk 来优化。

体积方面我们还没有超越 Webpack

**Reply by jinphic** [reply]
了解了，感谢解答

**Reply by nyngwang** [reply]
Does rspack have a plugin like the webpack counterpart to extract the runtime code for bundle size reduction?

**Reply by JSerFeng** [reply]
> Does rspack have a plugin like the webpack counterpart to extract the runtime code for bundle size reduction?

There is a option called optimization.runtimeChunk

**Comment by yanhaijing** [comment]
我测试的结果是rspack体积会大2%左右
